File: pytorch-examples/losses.py
import torch
import torch.nn.functional as F
from torch.optim import SGD


def main():
    # See the effect of a single step using SGD when directly maximizing a single class' probability
    logits = torch.tensor([[1.0, 1.0]], requires_grad=True)  # "parameters"
    y_true = torch.tensor([0])

    probs_before = torch.log_softmax(logits, dim=-1).exp()
    print(f"before: {logits.tolist()=}, {probs_before.tolist()=}")

    loss = F.nll_loss(logits, y_true)
    optim = SGD([logits])
    optim.zero_grad()
    loss.backward()
    optim.step()

    probs_after = torch.log_softmax(logits, dim=-1).exp()
    print(f"after: {logits.tolist()=}, {probs_after.tolist()=}")


if __name__ == "__main__":
    main()

File: examples/losses.rs
use crabgrad::engine::{exp, to_vec, DiscreteLabel, Value};
use crabgrad::nn::loss::{log_softmax, nll_loss_single};
use crabgrad::optim::{Optim, SGD};

fn main() {
    // Try once with ours
    let logits = vec![Value::from(1.0), Value::from(1.0)];
    let y_true = 0 as DiscreteLabel;

    let probs_before = exp(&log_softmax(&logits));
    dbg!("before", to_vec(&logits), to_vec(&probs_before));

    let loss = nll_loss_single(y_true, &log_softmax(&logits));
    let mut optim = SGD::new(&logits, 1e-3);
    optim.zero_grad();
    loss.backward();
    optim.step();

    let probs_after = exp(&log_softmax(&logits));
    dbg!("after", to_vec(&logits), to_vec(&probs_after));
}
